File: java/src/main/java/eu/bodynodesdev/common/BnUtils.java
/*
 * MIT License
 *
 * Copyright (c) 2025 Manuel Bottini
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS"; WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package eu.bodynodesdev.common;

public class BnUtils {

    // Prevent instantiation
    private BnUtils() {
        throw new UnsupportedOperationException("Class cannot be instantiated");
    }

    // Function to multiply two rotation matrices
    public static double[][] multiplyRotationMatrices(double[][] R1, double[][] R2) {
        return new double[][] {
            {
                R1[0][0] * R2[0][0] + R1[0][1] * R2[1][0] + R1[0][2] * R2[2][0],
                R1[0][0] * R2[0][1] + R1[0][1] * R2[1][1] + R1[0][2] * R2[2][1],
                R1[0][0] * R2[0][2] + R1[0][1] * R2[1][2] + R1[0][2] * R2[2][2]
            },
            {
                R1[1][0] * R2[0][0] + R1[1][1] * R2[1][0] + R1[1][2] * R2[2][0],
                R1[1][0] * R2[0][1] + R1[1][1] * R2[1][1] + R1[1][2] * R2[2][1],
                R1[1][0] * R2[0][2] + R1[1][1] * R2[1][2] + R1[1][2] * R2[2][2]
            },
            {
                R1[2][0] * R2[0][0] + R1[2][1] * R2[1][0] + R1[2][2] * R2[2][0],
                R1[2][0] * R2[0][1] + R1[2][1] * R2[1][1] + R1[2][2] * R2[2][1],
                R1[2][0] * R2[0][2] + R1[2][1] * R2[1][2] + R1[2][2] * R2[2][2]
            }
        };
    }

    // Function to multiply a rotation matrix and a vector
    public static double[] multiplyRotationMatrixWithVector(double[][] R1, double[] V1) {
        return new double[]
            {
                R1[0][0] * V1[0] + R1[0][1] * V1[1] + R1[0][2] * V1[2],
                R1[1][0] * V1[0] + R1[1][1] * V1[1] + R1[1][2] * V1[2],
                R1[2][0] * V1[0] + R1[2][1] * V1[1] + R1[2][2] * V1[2]
            };
    }

    // Function to compute a rotation matrix from Euler angles (XYZ order). Roll, pitch and yaw are in radians
    public static double[][] blender_euler_to_rotation_matrix_rad(double roll, double pitch, double yaw) {
        // counter rh wise rotations
        final double[][] R_x = new double[][]{
            { 1, 0, 0 },
            { 0, Math.cos(roll), -Math.sin(roll)},
            { 0, Math.sin(roll), Math.cos(roll) }
        };

        final double[][] R_y = new double[][]{
            { Math.cos(pitch), 0, Math.sin(pitch) },
            { 0, 1, 0 },
            { -Math.sin(pitch), 0, Math.cos(pitch) }
        };

        final double[][] R_z = new double[][]{
            {Math.cos(yaw), -Math.sin(yaw), 0},
            { Math.sin(yaw), Math.cos(yaw), 0},
            {0, 0, 1}
        };

        return multiplyRotationMatrices(R_z, multiplyRotationMatrices(R_y, R_x));
    }


    // Function to compute a rotation matrix from Euler angles (XYZ order). Roll, pitch and yaw are in degrees
    public static double[][] blender_euler_to_rotation_matrix_degree(double roll, double pitch, double yaw) {
        return blender_euler_to_rotation_matrix_rad(Math.toRadians(roll), Math.toRadians(pitch), Math.toRadians(yaw));
    }

    public static BnQuaternion createQuanternion(BnAxisConfig axis_config, double[] values) {
        return new BnQuaternion(
            axis_config.newWsign * values[axis_config.newWval],
            axis_config.newXsign * values[axis_config.newXval],
            axis_config.newYsign * values[axis_config.newYval],
            axis_config.newZsign * values[axis_config.newZval]
        );
    }

    /** This trasformation ignores what is the local axis of the object once rotated, we don't trust those axis system which might be compromised
    * sensorQuatVals               - is the raw quaternion (vector of 4 values) from the sensor
    * firstQuatVals                - [inout] is the first quaternion that has been registered from the sensor, this will create a 0 angle from where the sensor started sensing. Please fill it with 0,0,0,0 to set it as initially empty
    * startingQuatVals             - is the starting quaternion of the object we want to rotated with the sensor
    * envQuatVals                  - is the environment quaternion, basically indicates somehow where the x axis points 
    * pureAxisConfig               - axis configuration that will transfor the sensor values into values for the virtual world. Composed of 8 int array values that I can use to create a BnAxisConfig
    */
    public static double[] transformSensorQuat( double[] sensorQuatVals, double[] firstQuatVals, double[] startingQuatVals, double[] envQuatVals, int[] pureAxisConfig) {

        BnAxisConfig axisConfig = new BnAxisConfig(pureAxisConfig);

        BnQuaternion sensorQuat = new BnQuaternion(sensorQuatVals);
        BnQuaternion startingQuat = new BnQuaternion(startingQuatVals);
        BnQuaternion envQuat = new BnQuaternion(envQuatVals);

        BnQuaternion firstQuat = new BnQuaternion(firstQuatVals);
        if (firstQuat.isEmpty()) {
            firstQuat = sensorQuat.inverse();
            firstQuatVals[0] = firstQuat.getW();
            firstQuatVals[1] = firstQuat.getX();
            firstQuatVals[2] = firstQuat.getY();
            firstQuatVals[3] = firstQuat.getZ();
        }

        BnQuaternion rotationRealQuat = firstQuat.mul(sensorQuat);
        BnQuaternion rotationRealaxisQuat = createQuanternion(
                axisConfig,
                rotationRealQuat.toList());
        
        BnQuaternion objectNewQuat = envQuat.mul(rotationRealaxisQuat.mul(envQuat.inverse().mul(startingQuat)));
        
        return objectNewQuat.toList();
    }

}

File: cbasic/src/BnUtils.c
/*
 * MIT License
 *
 * Copyright (c) 2025 Manuel Bottini
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 all
 * copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS"; WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "BnUtils.h"
#include "math.h"

// Function to compute a rotation matrix from Euler angles (XYZ order). Roll,
// pitch and yaw are in radians
void BnUtils_blender_euler_to_rotation_matrix_rad(double roll, double pitch, double yaw, double (*const Ro)[3]) {
    // counter rh wise rotations
    double Rx[3][3] = {{1, 0, 0}, {0, cos(roll), -sin(roll)}, {0, sin(roll), cos(roll)}};

    double Ry[3][3] = {{cos(pitch), 0, sin(pitch)}, {0, 1, 0}, {-sin(pitch), 0, cos(pitch)}};

    double Rz[3][3] = {{cos(yaw), -sin(yaw), 0}, {sin(yaw), cos(yaw), 0}, {0, 0, 1}};

    double Rtmp[3][3];
    BnUtils_multiplyRotationMatrices(Ry, Rx, Rtmp);
    BnUtils_multiplyRotationMatrices(Rz, Rtmp, Ro);
}

void BnUtils_blender_euler_to_rotation_matrix_degree(double roll, double pitch, double yaw, double (*const Ro)[3]) {
    BnUtils_blender_euler_to_rotation_matrix_rad(BnUtils_toRadians(roll), BnUtils_toRadians(pitch),
                                                 BnUtils_toRadians(yaw), Ro);
}

BnQuaternion_t BnUtils_createQuanternion(BnAxisConfig_t const *const axis_config, double const *const values) {
    double valuesTmp[4] = {values[0], values[1], values[2], values[3]};
    BnAxisConfig_apply_double(axis_config, valuesTmp);
    return BnQuaternion_create_array(valuesTmp);
}

// all vals are double[4]
void BnUtils_transformSensorQuat(double const *const sensorQuatVals, double *const firstQuatVals,
                                 double const *const startingQuatVals, double const *const envQuatVals,
                                 int const *const pureAxisConfig, double *const valuesOut) {

    BnAxisConfig_t axisConfig;
    BnAxisConfig_config_purearray(&axisConfig, pureAxisConfig);

    BnQuaternion_t const sensorQuat = BnQuaternion_create_array(sensorQuatVals);
    BnQuaternion_t const startingQuat = BnQuaternion_create_array(startingQuatVals);
    BnQuaternion_t const envQuat = BnQuaternion_create_array(envQuatVals);

    BnQuaternion_t firstQuat = BnQuaternion_create_array(firstQuatVals);
    if (BnQuaternion_is_empty(&firstQuat) == 1) {
        firstQuat = BnQuaternion_inverse(&sensorQuat);
        BnQuaternion_to_list(&firstQuat, firstQuatVals);
    }

    BnQuaternion_t const rotationRealQuat = BnQuaternion_mul(&firstQuat, &sensorQuat);
    double valuesTmp[4];
    BnQuaternion_to_list(&rotationRealQuat, valuesTmp);
    BnQuaternion_t const rotationRealaxisQuat = BnUtils_createQuanternion(&axisConfig, valuesTmp);

    BnQuaternion_t objectNewQuat = BnQuaternion_inverse(&envQuat);
    objectNewQuat = BnQuaternion_mul(&objectNewQuat, &startingQuat);
    objectNewQuat = BnQuaternion_mul(&rotationRealaxisQuat, &objectNewQuat);
    objectNewQuat = BnQuaternion_mul(&envQuat, &objectNewQuat);
    BnQuaternion_to_list(&objectNewQuat, valuesOut);
}
